File: packages/python/goldenmatch/goldenmatch/semantic/metricflow.py
"""Read declared entity keys + measures from a dbt / MetricFlow semantic model.

MetricFlow declares, per `semantic_models[]`: `entities` (with `type` one of
primary/foreign/unique/natural), `measures`, and a default `agg_time_dimension`.
Joins across semantic models happen implicitly on *shared entities* — so the
primary entity's key is exactly what `certify_key_integrity` needs to check.

`parse_semantic_models(source) -> list[DeclaredKeySpec]` extracts that surface so
you can point the certifier at a real dbt project instead of hand-passing a key.
This reads the *declaration* only; it does not touch a warehouse.
"""
from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml

_PRIMARY_ENTITY_TYPES = {"primary", "natural"}


@dataclass
class DeclaredKeySpec:
    """One semantic model's declared identity, ready to feed `certify_key_integrity`."""

    model: str
    key: list[str]                       # primary/natural entity column(s)
    measures: list[str] = field(default_factory=list)
    grain: list[str] | None = None       # default agg_time_dimension, if any
    foreign_keys: list[str] = field(default_factory=list)  # foreign entities (join edges)


def _entity_column(entity: dict[str, Any]) -> str:
    """The physical column an entity maps to: `expr` if given, else `name`."""
    expr = entity.get("expr")
    if isinstance(expr, str) and expr.strip():
        return expr.strip()
    return str(entity.get("name", "")).strip()


def _measure_column(measure: dict[str, Any]) -> str:
    expr = measure.get("expr")
    if isinstance(expr, str) and expr.strip():
        return expr.strip()
    return str(measure.get("name", "")).strip()


def _load(source: str | Path | dict[str, Any]) -> dict[str, Any]:
    if isinstance(source, dict):
        return source
    if isinstance(source, (str, Path)) and os.path.exists(source):
        with open(source, encoding="utf-8") as fh:
            return yaml.safe_load(fh) or {}
    # treat as raw YAML text
    return yaml.safe_load(str(source)) or {}


def parse_semantic_models(source: str | Path | dict[str, Any]) -> list[DeclaredKeySpec]:
    """Parse dbt/MetricFlow `semantic_models` YAML into `DeclaredKeySpec`s.

    Args:
        source: a path to a YAML file, a raw YAML string, or an already-loaded dict.

    Returns:
        One `DeclaredKeySpec` per semantic model that declares a primary/natural
        entity (models without one are skipped — nothing to certify).
    """
    data = _load(source)
    models = data.get("semantic_models") or []
    specs: list[DeclaredKeySpec] = []

    for sm in models:
        if not isinstance(sm, dict):
            continue
        name = str(sm.get("name", "")).strip()
        entities = sm.get("entities") or []

        primary_key: list[str] = []
        foreign_keys: list[str] = []
        for ent in entities:
            if not isinstance(ent, dict):
                continue
            etype = str(ent.get("type", "")).strip().lower()
            col = _entity_column(ent)
            if not col:
                continue
            if etype in _PRIMARY_ENTITY_TYPES:
                primary_key.append(col)
            elif etype == "foreign":
                foreign_keys.append(col)
            elif etype == "unique":
                # a unique (non-primary) key is still a valid identity to check;
                # only promote it when no primary/natural entity was declared.
                foreign_keys.append(col)

        if not primary_key:
            # No primary/natural entity → nothing to certify for this model.
            continue

        measures = [
            c for m in (sm.get("measures") or [])
            if isinstance(m, dict) and (c := _measure_column(m))
        ]

        grain: list[str] | None = None
        agg_time = (sm.get("defaults") or {}).get("agg_time_dimension")
        if isinstance(agg_time, str) and agg_time.strip():
            grain = [agg_time.strip()]

        specs.append(
            DeclaredKeySpec(
                model=name,
                key=primary_key,
                measures=measures,
                grain=grain,
                foreign_keys=foreign_keys,
            )
        )

    return specs

File: packages/typescript/goldenmatch/src/core/semantic/metricflow.ts
/**
 * Emit dbt / MetricFlow semantic-model entity declarations (wedge B).
 *
 * Given a GoldenMatch-resolved key, generate the `semantic_models` YAML that
 * declares that conformed key as the PRIMARY entity — so every metric joins on
 * resolved identity. Faithful port of the emit half of Python
 * `semantic/metricflow.py`; the byte-for-byte YAML is produced by `dumpYaml`
 * (matching `yaml.safe_dump(sort_keys=False, default_flow_style=False)`).
 */

import { type YamlValue, dumpYaml } from "./yamlEmit.js";
import type { ResolvedCrosswalk } from "./crosswalk.js";
import { type LoadedDoc, asList, asStrStripped, isObj } from "./parseUtil.js";

const PRIMARY_ENTITY_TYPES = new Set(["primary", "natural"]);

/** One semantic model's declared identity, ready to feed `certifyKeyIntegrity`. */
export interface DeclaredKeySpec {
  model: string;
  key: string[]; // primary/natural entity column(s)
  measures: string[];
  grain: string[] | null; // default agg_time_dimension, if any
  foreignKeys: string[]; // foreign entities (join edges)
}

/** The physical column an entity maps to: `expr` if given, else `name`. */
function entityColumn(entity: Record<string, unknown>): string {
  const expr = entity["expr"];
  if (typeof expr === "string" && expr.trim()) return expr.trim();
  return asStrStripped(entity["name"]);
}

function measureColumn(measure: Record<string, unknown>): string {
  const expr = measure["expr"];
  if (typeof expr === "string" && expr.trim()) return expr.trim();
  return asStrStripped(measure["name"]);
}

/**
 * Parse dbt/MetricFlow `semantic_models` into `DeclaredKeySpec`s — one per model
 * that declares a primary/natural entity (models without one are skipped;
 * nothing to certify). Faithful port of Python `parse_semantic_models` (consume
 * half); operates on an already-loaded document object.
 */
export function parseSemanticModels(doc: LoadedDoc): DeclaredKeySpec[] {
  const specs: DeclaredKeySpec[] = [];
  for (const sm of asList(doc["semantic_models"])) {
    if (!isObj(sm)) continue;
    const name = asStrStripped(sm["name"]);

    const primaryKey: string[] = [];
    let foreignKeys: string[] = [];
    const uniqueKeys: string[] = [];
    for (const ent of asList(sm["entities"])) {
      if (!isObj(ent)) continue;
      const etype = asStrStripped(ent["type"]).toLowerCase();
      const col = entityColumn(ent);
      if (!col) continue;
      if (PRIMARY_ENTITY_TYPES.has(etype)) primaryKey.push(col);
      else if (etype === "foreign") foreignKeys.push(col);
      else if (etype === "unique") {
        uniqueKeys.push(col);
        // A unique (non-primary) entity is a join edge when a primary key is
        // present; it is promoted to the key only when no primary is.
        foreignKeys.push(col);
      }
    }

    const key = primaryKey.length ? primaryKey : uniqueKeys;
    if (!key.length) continue; // nothing to certify for this model
    if (!primaryKey.length) {
      // Promoted a unique key to the primary role — drop it from foreign_keys.
      foreignKeys = foreignKeys.filter((c) => !uniqueKeys.includes(c));
    }

    const measures: string[] = [];
    for (const m of asList(sm["measures"])) {
      if (!isObj(m)) continue;
      const c = measureColumn(m);
      if (c) measures.push(c);
    }

    let grain: string[] | null = null;
    const defaults = sm["defaults"];
    const aggTime = isObj(defaults) ? defaults["agg_time_dimension"] : undefined;
    if (typeof aggTime === "string" && aggTime.trim()) grain = [aggTime.trim()];

    specs.push({ model: name, key: [...key], measures, grain, foreignKeys });
  }
  return specs;
}

export interface EmitSemanticModelOptions {
  resolvedKey: string;
  entityName?: string;
  sourceKey?: string;
  measures?: readonly string[];
  grain?: readonly string[] | string | null;
  modelRef?: string;
  measureAgg?: string;
}

/**
 * Build ONE MetricFlow `semantic_models[]` entry declaring `resolvedKey` as the
 * PRIMARY entity — the GoldenMatch-conformed join key every metric inherits. The
 * original per-source key (`sourceKey`), if given and distinct, is declared a
 * `unique` entity so it stays queryable but is no longer the join primary.
 */
export function emitSemanticModel(
  model: string,
  opts: EmitSemanticModelOptions,
): { [k: string]: YamlValue } {
  const measureAgg = opts.measureAgg ?? "sum";
  const entities: YamlValue[] = [
    { name: opts.entityName || model, type: "primary", expr: opts.resolvedKey },
  ];
  if (opts.sourceKey && opts.sourceKey !== opts.resolvedKey) {
    entities.push({ name: opts.sourceKey, type: "unique", expr: opts.sourceKey });
  }

  const sm: { [k: string]: YamlValue } = {
    name: model,
    model: opts.modelRef || `ref('${model}')`,
    entities,
  };

  const grainList =
    typeof opts.grain === "string" ? [opts.grain] : [...(opts.grain ?? [])];
  if (grainList.length > 0) {
    sm["defaults"] = { agg_time_dimension: grainList[0]! };
  }

  const measures = opts.measures ?? [];
  if (measures.length > 0) {
    sm["measures"] = measures.map((m) => ({ name: m, agg: measureAgg, expr: m }));
  }
  return sm;
}

/**
 * Render one or more `emitSemanticModel` dicts as a `semantic_models` YAML
 * document (block style, key order preserved).
 */
export function emitMetricflowYaml(
  models: { [k: string]: YamlValue } | Array<{ [k: string]: YamlValue }>,
): string {
  const list = Array.isArray(models) ? models : [models];
  return dumpYaml({ semantic_models: list });
}

/**
 * Emit the `semantic_models` YAML for a `ResolvedCrosswalk`: its `resolvedKey`
 * becomes the primary entity, its `sourcePkColumn` the `unique` source key.
 */
export function emitFromCrosswalk(
  crosswalk: ResolvedCrosswalk,
  model: string,
  opts: {
    entityName?: string;
    measures?: readonly string[];
    grain?: readonly string[] | string | null;
    modelRef?: string;
  } = {},
): string {
  const sm = emitSemanticModel(model, {
    resolvedKey: crosswalk.resolvedKey ?? "resolved_entity_id",
    ...(opts.entityName !== undefined ? { entityName: opts.entityName } : {}),
    ...(crosswalk.sourcePkColumn !== undefined ? { sourceKey: crosswalk.sourcePkColumn } : {}),
    ...(opts.measures !== undefined ? { measures: opts.measures } : {}),
    ...(opts.grain !== undefined ? { grain: opts.grain } : {}),
    ...(opts.modelRef !== undefined ? { modelRef: opts.modelRef } : {}),
  });
  return emitMetricflowYaml(sm);
}
